### ISSUE TITLE
filter --trim true not working as advertised

### ISSUE THREAD

**Issue by dosumis** [ontodev/robot#473]
TL;DR: 
1. According to the doc:  `filter --term-file terms.txt --trim true` should return all axioms that only reference terms in terms.txt.  This is not the case.  I think it should be.
2. Better doc is needed on the interaction between filters specified with a list of terms and other filter options. It is not clear from doc what the result of `--trim true --select foo` will be for various flavours of foo (apart from "annotations, ontology").

Details:

"Each axiom refers to one or more entities: classes, object
properties, annotation properties, etc. *By default, filter will copy
axioms from the input ontology for which all the entities are in the
specified set.* To be more relaxed and copy axioms that contain one or
more entities from the specified set, use --trim false. This will
usually leave some “dangling references” to entities that were not in
the specified set."  [Robot Filter doc](http://robot.obolibrary.org/filter)
[cl-int.owl.zip](https://github.com/ontodev/robot/files/3134045/cl-int.owl.zip)
[terms.txt](https://github.com/ontodev/robot/files/3134048/terms.txt)

### Example

cl-int.owl (attached) term.txt(attached) - contains all CL terms + two object properties, one of which is RO_0002202

cl_int.owl has:

```xml
    <owl:Class rdf:about="http://purl.obolibrary.org/obo/CL_0000017">
        <rdfs:subClassOf rdf:resource="http://purl.obolibrary.org/obo/CL_0000015"/>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://purl.obolibrary.org/obo/RO_0002202"/>
                <owl:someValuesFrom rdf:resource="http://purl.obolibrary.org/obo/CL_0000020"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://purl.obolibrary.org/obo/RO_0002215"/>
                <owl:someValuesFrom rdf:resource="http://purl.obolibrary.org/obo/GO_0048137"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <obo:IAO_0000115 rdf:datatype="http://www.w3.org/2001/XMLSchema#string">A male germ cell that develops from spermatogonia. The euploid primary spermatocytes undergo meiosis and give rise to the haploid secondary spermatocytes which in turn give rise to spermatids.</obo:IAO_0000115>

```

     robot filter -i cl-int.owl --term-file terms.txt \ 
      --trim false --select annotations --preserve-structure false --output \ 
     cl_filter_trim_false.owl

=> 

```xml
    <!-- http://purl.obolibrary.org/obo/CL_0000017 -->

    <owl:Class rdf:about="http://purl.obolibrary.org/obo/CL_0000017">
        <rdfs:subClassOf rdf:resource="http://purl.obolibrary.org/obo/CL_0000015"/>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://purl.obolibrary.org/obo/RO_0002202"/>
                <owl:someValuesFrom rdf:resource="http://purl.obolibrary.org/obo/CL_0000020"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <rdfs:subClassOf>
            <owl:Restriction>
                <owl:onProperty rdf:resource="http://purl.obolibrary.org/obo/RO_0002215"/>
                <owl:someValuesFrom rdf:resource="http://purl.obolibrary.org/obo/GO_0048137"/>
            </owl:Restriction>
        </rdfs:subClassOf>
        <obo:IAO_0000115 rdf:datatype="http://www.w3.org/2001/XMLSchema#string">A male germ cell that develops from spermatogonia. The euploid primary spermatocytes undergo meiosis and give rise to the haploid secondary spermatocytes which in turn give rise to spermatids.</obo:IAO_0000115>
 
```
As expected GO_0048137 is retained as a dangling ref.  However

     robot filter -i cl-int.owl --term-file terms.txt \
        --trim true --preserve-structure false --output cl_filter_trim_true_na.owl

=> 

```xml
      <!-- http://purl.obolibrary.org/obo/CL_0000017 -->

    <owl:Class rdf:about="http://purl.obolibrary.org/obo/CL_0000017">
        <rdfs:subClassOf rdf:resource="http://purl.obolibrary.org/obo/CL_0000015"/>
    </owl:Class>

```

This is rather more surprising. CL:0000017, RO:0002202 and CL:0000020
are all referenced in ontologyterms.txt, but the subClassOf
restriction referncing all 3 is not retained.  In contrast CL:0000017
subClassOf CL:0000015, for which both terms are referenced, is retained.  Surely this is a bug.  Having this work as advertised would both useful in itself and improve usability (Users wouldn't need to spend time experimenting with combining --select options to try to get what  they want).

The only way I've found  to get ~what I want is:

    robot  filter -i cl-int.owl --term-file terms.txt --trim true \
       --select "annotations anonymous parent" --preserve-structure false \
       --output cl_filter_trim_true_select_ann_anon_parent.owl

The problem with this is that I pull back undesired terms that are not in the filter but that are parents of the specified filter terms.

The behavior of --select when combined with --trim true should be better documented.  Is it accurate to say the --select selectively eliminates trimming of axioms?  So, `--select annotations` eliminates trimming of annotation axioms on selected terms (& axioms?).  `--select parents` selectively eliminates trimming of asserted parent axioms etc?  Or do some of these --select options add additional terms to the filter set?

**Comment by jamesaoverton** [comment]
Hi @dosumis, thanks for the detailed report. @rctauber and I are seeing what you describe.

This is something we should be able to do, but I think you hit a limitation of the current implementation. Let's try to figure this out.

Yes, `--select` modifies the set of target entities. Then `--trim` says whether we care about "complete" axioms where all the entities are in the target set, or "partial" axioms where one or more entities is in the target set.

I think that the problem in this case is that the selectors are either including too much or too little, and we aren't providing a selector for exactly what you want. This is the relevant axiom in RDFXML:

```
<rdfs:subClassOf>
    <owl:Restriction>
        <owl:onProperty rdf:resource="http://purl.obolibrary.org/obo/RO_0002202"/>
        <owl:someValuesFrom rdf:resource="http://purl.obolibrary.org/obo/CL_0000020"/>
    </owl:Restriction>
</rdfs:subClassOf>
```

and in OWL Functional Syntax:

```
SubClassOf(<http://purl.obolibrary.org/obo/CL_0000017> ObjectSomeValuesFrom(<http://purl.obolibrary.org/obo/RO_0002202> <http://purl.obolibrary.org/obo/CL_0000020>)))
```

from which we generate this set of four elements:

```
[<http://purl.obolibrary.org/obo/CL_0000017>,
 ObjectSomeValuesFrom(<http://purl.obolibrary.org/obo/RO_0002202> <http://purl.obolibrary.org/obo/CL_0000020>), 
 <http://purl.obolibrary.org/obo/RO_0002202>,
 <http://purl.obolibrary.org/obo/CL_0000020>]
```

When `--trim false`, we include this axiom in the results because at least one of the elements of the axiom set is in the target set you specified: the three named classes. When `--trim true` we do not include this axiom because not all of the elements are in your target set: the anonymous `ObjectSomeValuesFrom` is not. (There are use cases where distinguishing named an anonymous entities is important.)

When you `--select "self annotations parents"` then the anonymous `ObjectSomeValuesFrom` is added to your target set, along with other stuff. Now when `--trim true` then all four elements are in the target set, and the axiom is included. However it sounds like this is including parents that you don't want.

Right now I'm thinking that an `anonymous-parents` selector would do the trick, but I want to give that some more thought.

We tried not to get into this depth of technical details in the docs, but maybe more detail is needed.

**Comment by dosumis** [comment]
Perhaps it's just my perspective, but the current behavior feels overly RDF-centric with the result that it is inevitably hard to use for those of us used to thinking at the level of OWL abstraction.  I read 'entity' in the doc as owl entity (class, individual,  objectProperty, dataProperty, annotationProperty).   Having to find the right combination of (poorly documented) --SELECT options to achieve something as simple as returning all axioms with specified owl entities makes using ROBOT filters quite unituitive when working with OWL.

Would it be possible to have the default behavior of --trim true be that OWL keywords are ignored  and to support the more basal RDF use-cases with additional options?   I  suspect this will make filtering much more usable for the majority of robot users. 






  At the level of OWL abstraction it seems odd to treat `ObjectSomeValuesFrom` as an entity  (it's not an OWLentity type).  From an OWL perspective,  what I'm trying to  do is simple:  I want all the axioms (of any type) that only reference a specified set of OWL entities.  Having  to work out which obscure --select option to choose to achieve this means `robot filter` is not easily usable for OWL-centric work.

I agree it is useful to

**Comment by dosumis** [comment]
(Apologies for rough notes at bottom of last comment, now removed)

**Comment by jamesaoverton** [comment]
I recognize the specific use case you've described, and I acknowledge that it's something that I want ROBOT to support in a straightforward way. I'm not clear on your general goal with this CL example, so maybe there's better way to achieve that goal with other ROBOT commands, such as `extract`.

The `remove` and `filter` commands perform low-level delete/copy operations on individual OWL axioms. They operate exclusively on `OWLObject`s (named and anonymous) and `OWLAxiom`s, not RDF. They use only OWLAPI methods. Almost all the heavy lifting for `--select` is done by OWLAPI's `org.semanticweb.owlapi.search.EntitySearcher`, as you can see in [`RelatedObjectsHelper.java`](https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java). So the design is entirely driven by OWL and existing OWLAPI methods, and deliberately at a low level.

I'm not particularly happy with how `--trim` has turned out, which is part of the problem in this case, and I'll give it more thought.

**Comment by dosumis** [comment]
Aim: return all axioms with specified owl entities.

>  When --trim true we do not include this axiom because not all of the elements are in your target set: the anonymous ObjectSomeValuesFrom is not. 

Given this, shouldn't adding `--strict true --select anonymous` be sufficient to return all anonymous axioms that reference only terms in the reference list?  It currently is not.

**Comment by dosumis** [comment]
> I'm not clear on your general goal with this CL example, so maybe there's better way to achieve that goal with other ROBOT commands, such as extract.

See "Release artefact 4: simple" section here: https://github.com/INCATools/ontology-development-kit/issues/194

**Comment by jamesaoverton** [comment]
Ok. The release artifacts are an important use case, no question. I'll assign this to @rctauber and we'll figure out how to get the desired behaviour, with or without `filter`. I'd also be fine with a specialized ROBOT command for generating these artefacts.

Help with detailed use cases, examples, and code is always appreciated.

**Comment by dosumis** [comment]
Reminder: This is a major blocker for the ODK.  Because of it, we've had to revert some pipelines to using OWLtools & OORT in order to make OBO output files for some projects.

Any chance of a fix soon?

**Comment by jamesaoverton** [comment]
We proposed a solution #483 with code #484 a month ago.

We need a known-good example of some ontology and its `simple` form, that we can test. I looked around GitHub, and although I can see `*-simple.owl` in several Makefiles, and I see how `owltools` is called, I didn't find a OWL file that I can just download and test. Can you please point me to a known-good test case?

**Comment by matentzn** [comment]
Hey @jamesaoverton 

Sorry for the delay, David is travelling around but I will have him comment when he is back on Thursday! :) Have a great one! BTW: When is the next release planned? The next ODK update needs the fixes and extensions you have implemented since the last release :P 

I do think so that this here issue should be addressed before release, so I will make sure you will have a comment on thursday. Thanks!

**Comment by cmungall** [comment]
The spec for obo-basic is here: #17

**Comment by jamesaoverton** [comment]
@cmungall In https://github.com/INCATools/ontology-development-kit/issues/194 we have "Release artefact 4: simple (optional)" and "Release artefact 5: basic". Issue #17 says "basic" but seems to be referring to "simple". This is confusing.

@matentzn: This issue is important and I don't want to rush into the wrong resolution. So I can make a new release now (after merging the `template` PR) or I can wait for this. Let me know what you prefer.

**Comment by matentzn** [comment]
Ok, I personally favour Friday to make a decision; and maybe a new release next week? :P

**Comment by matentzn** [comment]
Another simple example:

ont.owl:
```
Prefix(:=<http://purl.obolibrary.org/obo/wbbt.owl#>)
Ontology(<http://purl.obolibrary.org/obo/wbbt.owl>
<http://purl.obolibrary.org/obo/wbbt/2018-10-19/wbbt.owl>
EquivalentClasses(<http://purl.obolibrary.org/obo/WBbt_0008452> ObjectUnionOf(<http://purl.obolibrary.org/obo/WBbt_0002463> <http://purl.obolibrary.org/obo/WBbt_0003254>))
)
```

seed.txt:
```
http://purl.obolibrary.org/obo/WBbt_0008452
http://purl.obolibrary.org/obo/WBbt_0002463
http://purl.obolibrary.org/obo/WBbt_0003254
```

Running
```
robot filter --input ont.owl --term-file seed.txt --trim true --select "anonymous parents object-properties self" --output out.owl
```

should afaic understand from the thread here not remove the axiom, but does.

**Comment by jamesaoverton** [comment]
Thanks for the test case. `out.owl` is empty because the `ObjectUnionOf` is anonymous and the selectors are not reaching it: it's anonymous but it's not part of the set of three terms that `--select` is acting on, so the `anonymous` selector doesn't grab it; it's not in `parents`; it's not in `object-properties`; and it's not in `self`, which is the set of three terms from `seed.txt`. Since the target set does not include all objects in the axiom, the axiom is not copied to `out.owl`.

Reviewing the `remove` documentation again, I now think this part is misleading: "By default, both `named` and `anonymous` entities are removed. But, for example, if only `--select anonymous` is provided, the named classes will not be removed."

Using `robot filter --input ont.owl --term-file seed.txt --trim true --select "self equivalents" --output out.owl` includes the axiom in `out.owl`: `self` includes the three original terms; `equivalents` includes the anonymous `ObjectUnionOf`; putting them together in one `--select` takes the union of those two; since the target set now includes all object in the axiom, the axiom is copied to `out.owl`. But I don't think that adding the `equivalents` selector addresses the original issue.

Using #484 `robot filter --input ont.owl --term-file seed.txt --trim true --signature true --output out.owl` includes the axiom in `out.owl`, because only named OWLEntities are considered. I think that addresses the original issue.

**Comment by jamesaoverton** [comment]
I just released [1.4.1](https://github.com/ontodev/robot/releases/tag/v1.4.1), which I hope addresses this.

**Comment by matentzn** [comment]
Amazing James! I will get right on to test this next week! Thaaanks so much!

**Comment by beckyjackson** [comment]
I think release 1.4.1 resolved this, as @jamesaoverton said above. Is this OK to close?

**Comment by matentzn** [comment]
I will re-open in case this is still an issue. I did not notice anything so far!
